arcgis-ios-sdk-samples/Maps/Show location history/LocationTracking.swift

```diff
@@ -0,0 +1,185 @@
+        
+        let trackSymbol = AGSSimpleLineSymbol(style: .solid, color: Constants.trackColor, width: Constants.trackWidth)
+        overlay.renderer = AGSSimpleRenderer(symbol: trackSymbol)
+        
+        return overlay
+    }()
+    
+    private var trackBuilder: AGSPolylineBuilder?
+    
+    // MARK: Internal behavior
+    
+    init(mapView: AGSMapView, historyView: LocationHistoryView, dataSource: AGSLocationDataSource? = nil) {
+        self.mapView = mapView
+        self.historyView = historyView
+        self.dataSource = dataSource
+        
+        setupMapView(mapView)
+    }
+    
+    func toggleTrackingStatus() {
+        switch currentStatus {
+        case .paused:
+            currentStatus = .inProgress
+        case .inProgress:
+            currentStatus = .paused
+        }
+    }
+    
+    // MARK: Private behavior
+    
+    private func handleLocationStatusChange() {
+        switch currentStatus {
+        case .paused:
+            stopProcessingLocationChanges()
+        case .inProgress:
+            startProcessingLocationChanges()
+        }
+        
+        updateView(status: currentStatus)
+    }
+    
+    private func processLocationUpdate() {
+        guard currentStatus == .inProgress, let position = mapView?.locationDisplay.mapLocation, position.x != 0, position.y != 0 else { return }
+
+        let locationGraphic = AGSGraphic(geometry: position, symbol: nil, attributes: nil)
+        locationsOverlay.graphics.add(locationGraphic)
+
+        guard let trackBuilder = trackBuilder else { return }
+
+        trackBuilder.add(position)
+        let trackGraphic = AGSGraphic(geometry: trackBuilder.toGeometry(), symbol: nil, attributes: nil)
+        trackOverlay.graphics.add(trackGraphic)
+    }
+    
+    private func startProcessingLocationChanges() {
+        mapView?.locationDisplay.locationChangedHandler = { [weak self] (location) in
+            guard location.horizontalAccuracy >= 0 else { return }
+            
+            DispatchQueue.main.async {
+                self?.processLocationUpdate()
+            }
+        }
+    }
+    
+    private func stopProcessingLocationChanges() {
+        mapView?.locationDisplay.locationChangedHandler = nil
+    }
+    
+    private func setupLocationDisplay(_ locationDisplay: AGSLocationDisplay) {
+        if let specifiedDataSource = dataSource {
+            locationDisplay.dataSource = specifiedDataSource
+        }
+        
+        locationDisplay.autoPanMode = .recenter
+        locationDisplay.initialZoomScale = Constants.initialZoomScale
+        
+        locationDisplay.start { (error) in
+            if let error = error {
+                print("Error starting location display: \(error.localizedDescription)")
+            }
+        }
+    }
+    
+    private func setupMapView(_ mapView: AGSMapView) {
+        mapView.map = map
+        
+        map.load(completion: { [weak self, loadedMap = self.map] (error) in
```

Comment:
It was a good idea to maintain a reference to the map instead of going through `self` inside the closure. Unfortunately, `loadedMap` is the thing being loaded and it maintains a strong reference to the completion handler. To prevent a retain cycle between the map and the view controller, please use an `unowned` reference:

```swift
map.load { [weak self, unowned map] (error) in
```

Reply:
Good catch - thank you!

Reply:
I have addressed this via [`e367330`](https://github.com/Esri/arcgis-runtime-samples-ios/pull/751/commits/e3673303e3e56a88f7d7197bd43442891f8e0169).